File: server/src/archives/metadata/ccdc06.rs
use super::{ArchiveMetadata, MultiIdField, MultiTextField};
use crate::{archives, utils};
use serde::Deserialize;
use std::collections::HashMap;

#[derive(Deserialize)]
pub struct Metadata {
  #[serde(rename = "Title")]
  pub title: String,
  #[serde(rename = "Description", default)]
  pub description: Option<String>,
  #[serde(rename = "Artist", default)]
  pub artists: Option<MultiTextField>,
  #[serde(rename = "Circle", default)]
  pub circles: Option<MultiTextField>,
  #[serde(rename = "Magazine", default)]
  pub magazines: Option<MultiTextField>,
  #[serde(rename = "Event", default)]
  pub events: Option<MultiTextField>,
  #[serde(rename = "Publisher", default)]
  pub publishers: Option<MultiTextField>,
  #[serde(rename = "Parody", default)]
  pub parodies: Option<MultiTextField>,
  #[serde(rename = "Tags", default)]
  pub tags: Option<Vec<String>>,
  #[serde(rename = "URL", default)]
  pub urls: Option<HashMap<String, String>>,
  #[serde(rename = "Id", default)]
  pub source_ids: Option<HashMap<String, MultiIdField>>,
  #[serde(rename = "Released", default)]
  pub released: Option<i64>,
  #[serde(rename = "ThumbnailIndex", default)]
  pub thumb_index: Option<i16>,
  #[serde(rename = "Files", default)]
  pub files: Option<Vec<String>>,
}

pub fn add_metadata(info: Metadata) -> ArchiveMetadata {
  let mut archive = ArchiveMetadata {
    title: info.title,
    description: info.description,
    thumbnail: info.thumb_index.map(|index| index + 1),
    released_at: utils::map_timestamp(info.released),
    artists: info.artists.map(|field| field.to_vec()),
    circles: info.circles.map(|field| field.to_vec()),
    magazines: info.magazines.map(|field| field.to_vec()),
    events: info.events.map(|field| field.to_vec()),
    publishers: info.publishers.map(|field| field.to_vec()),
    parodies: info.parodies.map(|field| field.to_vec()),
    tags: info
      .tags
      .map(|tags| tags.into_iter().map(|tag| (tag, "".to_string())).collect()),
    ..Default::default()
  };

  let mut sources: Vec<archives::Source> = vec![];

  if let Some(urls) = info.urls {
    for (name, url) in urls {
      sources.push(archives::Source {
        name: utils::parse_source_name(&name),
        url: Some(url),
      })
    }
  }

  if let Some(source_ids) = info.source_ids {
    for (name, id) in source_ids {
      sources.push(archives::Source {
        name: utils::parse_source_name(&name),
        url: parse_source_id(name, id),
      })
    }
  }

  archive.sources = Some(sources);

  archive.images = info.files.map(|files| {
    files
      .iter()
      .enumerate()
      .map(|(i, file)| archives::Image {
        filename: file.to_string(),
        page_number: (i + 1) as i16,
        ..Default::default()
      })
      .collect()
  });

  archive
}

fn parse_source_id(name: String, id: MultiIdField) -> Option<String> {
  match name.to_lowercase().as_str() {
    "anchira" => Some(format!("https://anchira.to/g/{id}")),
    "hentainexus" => Some(format!("https://hentainexus.com/view/{id}")),
    _ => None,
  }
}

File: cli/metadata/ccdc06.ts
import dayjs from 'dayjs';
import YAML from 'yaml';
import { z } from 'zod';

import { ArchiveMetadata } from '../../shared/metadata';
import { mapMultiField, multiTextField } from './schemas';

const multiIdField = z.union([z.number(), z.string()]);

type MultiIdField = z.infer<typeof multiIdField>;

const parseSourceId = (name: string, id: MultiIdField) => {
	switch (name.toLowerCase()) {
		case 'anchira':
			return `https://anchira.to${id}`;
		case 'hentainexus':
			return `https://hentainexus.com/view/${id}`;
		case 'koharu':
			return `https://koharu.to${id}`;
	}
};

const metadataSchema = z.object({
	Title: z.string(),
	Description: z.string().optional(),
	Artist: multiTextField.optional(),
	Circle: multiTextField.optional(),
	Magazine: multiTextField.optional(),
	Event: multiTextField.optional(),
	Publisher: multiTextField.optional(),
	Parody: multiTextField.optional(),
	Tags: z.array(z.string()).optional(),
	URL: multiTextField.optional(),
	Id: z.record(z.string(), multiIdField).optional(),
	Released: z.number().optional(),
	ThumbnailIndex: z.number().optional(),
	Files: z.array(z.string()).optional(),
});

export default async (content: string, archive: ArchiveMetadata) => {
	const parsed = YAML.parse(content);
	const { data, error } = metadataSchema.safeParse(parsed);

	if (!data) {
		throw new Error(`Failed to parse CCDC06 metadata: ${error}`);
	}

	archive = structuredClone(archive);

	archive.title = data.Title;
	archive.description = data.Description;
	archive.thumbnail = data.ThumbnailIndex !== undefined ? data.ThumbnailIndex + 1 : undefined;
	archive.releasedAt = data.Released ? dayjs.unix(data.Released).toDate() : undefined;

	archive.tags = [];

	archive.tags.push(
		...mapMultiField(data.Artist).map((tag) => ({ namespace: 'artist', name: tag }))
	);
	archive.tags.push(
		...mapMultiField(data.Circle).map((tag) => ({ namespace: 'circle', name: tag }))
	);
	archive.tags.push(
		...mapMultiField(data.Magazine).map((tag) => ({ namespace: 'magazine', name: tag }))
	);
	archive.tags.push(...mapMultiField(data.Event).map((tag) => ({ namespace: 'event', name: tag })));
	archive.tags.push(
		...mapMultiField(data.Publisher).map((tag) => ({ namespace: 'publisher', name: tag }))
	);
	archive.tags.push(
		...mapMultiField(data.Parody).map((tag) => ({ namespace: 'parody', name: tag }))
	);

	if (data.Tags) {
		archive.tags.push(...data.Tags.map((tag) => ({ namespace: 'tag', name: tag })));
	}

	archive.sources = [];

	for (const url of mapMultiField(data.URL)) {
		archive.sources.push({ url });
	}

	if (data.Id) {
		for (const [name, id] of Object.entries(data.Id)) {
			archive.sources.push({
				name,
				url: parseSourceId(name, id),
			});
		}
	}

	archive.imageOrder = data.Files?.map((filename, i) => ({
		filename,
		pageNumber: i + 1,
	}));

	return archive;
};
